src/k8s_upgrade.py

```diff
@@ -0,0 +1,405 @@
+            charm.on["force-refresh-start"].action, self._on_force_upgrade_action
+        )
+
+    def _on_force_upgrade_action(self, event: ActionEvent):
+        if not self.charm.unit.is_leader():
+            message = f"Must run action on leader unit. (e.g. `juju run {self.charm.app.name}/leader force-refresh-start`)"
+            logger.debug(f"Force refresh failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade or not self._upgrade.in_progress:
+            message = "No upgrade in progress"
+            logger.debug(f"Force refresh failed: {message}")
+            event.fail(message)
+            return
+        self._upgrade.reconcile_partition(action_event=event)
+
+    def _reconcile_upgrade(self, event: EventBase) -> None:
+        """Handle upgrade events."""
+        if not self._upgrade:
+            logger.debug("Peer relation not available")
+            return
+        if not self._upgrade.versions_set:
+            logger.debug("Peer relation not ready")
+            return
+        if self.charm.unit.is_leader() and not self._upgrade.in_progress:
+            # Run before checking `self._upgrade.is_compatible` in case incompatible upgrade was
+            # forced & completed on all units.
+            self._upgrade.set_versions_in_app_databag()
+
+        if self._upgrade.unit_state is UnitState.RESTARTING:  # Kubernetes only
+            if not self._upgrade.is_compatible:
+                logger.info(
+                    "Refresh incompatible. If you accept potential *data loss* and *downtime*, you can continue with `force-refresh-start`"
+                )
+                self.charm.status.set_and_share_status(Config.Status.UNHEALTHY_UPGRADE)
+                return
+        if self.charm.db_initialised and self.charm.is_db_service_ready():
+            self._upgrade.unit_state = UnitState.HEALTHY
+        if self.charm.unit.is_leader():
+            self._upgrade.reconcile_partition()
+
+        self._set_upgrade_status()
+
+    def _set_upgrade_status(self):
+        if self.charm.unit.is_leader():
+            self.charm.app.status = self._upgrade.app_status or ActiveStatus()
+        # Set/clear upgrade unit status if no other unit status - upgrade status for units should
+        # have the lowest priority.
+        if isinstance(self.charm.unit.status, ActiveStatus) or (
+            isinstance(self.charm.unit.status, BlockedStatus)
+            and self.charm.unit.status.message.startswith(
+                "Rollback with `juju refresh`. Pre-refresh check failed:"
+            )
+        ):
+            self.charm.status.set_and_share_status(
+                self._upgrade.get_unit_juju_status() or ActiveStatus()
+            )
+
+    def _on_upgrade_peer_relation_created(self, _) -> None:
+        if self.charm.unit.is_leader():
+            self._upgrade.set_versions_in_app_databag()
+
+    def run_post_app_upgrade_task(self, event: EventBase):
```

Comment:
The big difference here is the `_determine_partition` function: In the mongos charm we don't have a `resume-refresh` action so unless the unit fails to upgrade, it will continue